### PR TITLE
Feature: Add Google Cloud Storage to Samba operator

### PR DESCRIPTION
**Pull request by ealebed**
---

Added new Google Cloud Storage to Samba server operator

### CODE REVIEW

File: airflow/providers/google/cloud/transfers/gcs_to_samba.py
Comment (eladkal) [inline airflow/providers/google/cloud/transfers/gcs_to_samba.py:1]: some might disagree but I would claim that samba is the higher intrest provider here
https://github.com/apache/airflow/blob/cf0c8245f54349fdcfe9ecdc19b34d8c71cfd479/CONTRIBUTING.rst#L1123-L1129

so I believe this file should be in samba provider and not in google

Comment (ealebed) [reply airflow/providers/google/cloud/transfers/gcs_to_samba.py:1]: Thanks @eladkal, you are absolutely right!
I've just moved operator to the correct folder, has highest "interest" in the transfer operator

Comment (potiuk) [reply airflow/providers/google/cloud/transfers/gcs_to_samba.py:1]: > some might disagree but I would claim that samba is the higher intrest provider here

I'd disagree actually :) It's a bit blurry of course, but I think our rules are rather clear that the "target" STAKEHOLDER interest is what matters when we make the decision. But in order to have the interest, it would have have to have a stakeholder. This is mainly the question of "who will likely want to maintain the transfer operator because they want your data in".

With services like GCP/AWS -> it's very clear who the stakeholder is and why they are interested in having transfers to them.  For example if you derive a new compression and make it faster to upload something to GCS, it's clear that Google team will be interested in adding the option to use it.

With SAMBA it's much less clear who the stakeholder is. Samba is a program that implements CIFS protocol https://www.samba.org/samba/docs/using_samba/ch01.html - it's not a service. Those who release samba are not interested in getting the data from GCS. They are indifferent about that to be honest. When you use GCS to  SAMBA, the other end of the transfer is ... YOU - the user who has a Windows network drive. So what happens in this case when you have GCS -> SAMBA you only have one stakeholder (Google Cloud). The other stakeholder is missing.

We have a number of similar cases (for example GCSToLocalFilesystemOperator) which is placed in Google provider. This is a very similar story. I think this one should also be in `google` provider.

Comment (eladkal) [reply airflow/providers/google/cloud/transfers/gcs_to_samba.py:1]: I look at this very differently.
For me the most important thing is **user comfort**. Where users are likely to look for this integration. It should be simple and clear. We should not have completed rules.

The shared goverance model does not bind google stackholders to contribute only to goolge provider. They can have intrest also in other providers. If Samba provider will not work then the google integration wont work regardless of where the class is located.


I do think that we should finalize the policy around this cases. The blury rules make it so that we are challanging the policy with every PR.
I'd be happy to start a mailing list thread after the summit.

Comment (potiuk) [reply airflow/providers/google/cloud/transfers/gcs_to_samba.py:1]: > For me the most important thing is user comfort. Where users are likely to look for this integration. It should be simple and clear. We should not have completed rules.

I am not sure how we assess that it's neither simple nor clear and it refers to very subjective assesment of "where the user will look for something when there are two ends". 

I am not sure where users will be likely to look for this integration - in GCP ? in Samba?. No idea. When the user wants to use 5-10 GCS releated transfers, they will look in 'google". When the user only has samba and wants to look for all the possible ways Samba can interact with external world, they will look in Samba. But most likely the user will look for 'How should I move from GCS To Samba or from Samba to GCS" and will look in both.

But regardless, if we are looking to help users to find something, maybe better solution will be to have a simple search on our website that will let them find it. The attempts to foresee in which directories/indexes people will search for something was simply not working - simple search where you enter something you look for is better. Yahoo and Altavista story vs. Google have proven that for a long time. I think generally no-one looks for things by traversing tree-like-hierarchical indexes. That's why in this case I think optimising for maintenance is better then optimising for (doubtful) search convenience. But maybe I am wrong of course.

There is no "good" solution for transfer operators. There are only trade-offs. And I think maintenance likelihood as an optimisation goal is good one.

> I do think that we should finalize the policy around this cases. The blury rules make it so that we are challanging the policy with every PR.

I would not exaggerate, It's Just a few PRs. And mostly the two of us argue about it :)

> The shared goverance model does not bind google stackholders to contribute only to goolge provider. They can have intrest also in other providers. 

Not at all. It's quantum more than 0/1 logic. It's all about higher change of being interested in making it works.

> If Samba provider will not work then the google integration wont work regardless of where the class is located.

Not really. Samba in this case is an optional feature of google provider  (it will be a [samba] extra) and until you import this particular class it won't even import samba classes. 

> I'd be happy to start a mailing list thread after the summit.

Sure. No problems. But I think it's just wrong optimisation goal. Also before you send it - I suggest you make a complete proposal and see what you come up with - i.e. take the transfers we have and map them according to rules you propose. You might be surprised how "unobvious" this will look. But yeah. I might be wrong of course.

Comment (potiuk) [reply airflow/providers/google/cloud/transfers/gcs_to_samba.py:1]: Also one more thought that this discussion inspired (and maybe you should consider that as an option in your proposal). If you want to also optimise for findability - why don't we simply import the Google transfer operator In the samba provider in transfer? Just import It there In similarly named package. (And same for all other transfers) 

This seems to solve the problem you mentioned -of finding the operator  - user will FIND and will be able to use it in both providers.  They will be able to find them in both places in docs and import from whatever provider they choose to. It will work.


That will remove the user out of the discussion because then it won't matter for the user. It will leave only the maintainers in the discussion and there I believe the currently agrees definition is good.

**Comment by ealebed** [comment]
Hi @eladkal 👋 
Could you please describe what should I do next? Should I move this operator into another location or not?

File: airflow/providers/google/cloud/transfers/gcs_to_samba.py
Comment (eladkal) [reply airflow/providers/google/cloud/transfers/gcs_to_samba.py:1]: I don't want to block this PR because the policy is blurry.
I think we can accept it as is for now (without this being a precedent). I will issue a mailing list thread and we will get the policy cleared, then we will adjust the code as needed.

**Review by eladkal — APPROVED** [review]

**Pull request merged**
